munit/shared/src/main/scala/munit/FunFixtures.scala

```diff
@@ -1,29 +1,56 @@
 package munit
 
+import munit.internal.FutureCompat._
+
+import scala.concurrent.Future
+
 trait FunFixtures { self: FunSuite =>
 
-  class FunFixture[T](
-      val setup: TestOptions => T,
-      val teardown: T => Unit
+  class FunFixture[T] private (
+      val setup: TestOptions => Future[T],
+      val teardown: T => Future[Unit]
   ) {
```

Comment:
That is actually a great idea! I have done this many times, and I am not sure how this escaped my mind this time. :)

Reply:
This turned out to be a bit harder than initially thought. It is not possible to have such constructor override, because erased types of both constructors become the same.

So I resorted to changing the class val to a `Tuple` so we can have the old constructor around and deprecate it. Once it is gone, we can switch back to the previous class vals.

Reply:
You can work around erasure conflicts by adding an `implicit dummy: DummyImplicit` parameter. The following program compiles OK
```scala
class A private (fn: Int => String)(implicit dummy: DummyImplicit) {
  def this(fn: Int => List[Int]) = this(n => fn(n).toString)
}
```

Reply:
Aha! I have seen this trick before somewhere. Will try!